editoast/core_task/src/envs/pathfinding.rs

```diff
@@ -337,44 +278,127 @@ where
             .collect()
     }
 
-    /// Builds the [Self::trains] map using trains from both [Self::consists] and [Self::constraints]
-    ///
-    /// TODO: we should build it incrementally to avoid unnecessary complexity
-    fn build_input_map(&mut self) {
-        self.trains.clear();
-        for train in self.all_trains() {
-            let consist = self.consists.get(&train).expect("all_trains invariant");
-            let constraints = self.constraints.get(&train).expect("all_trains invariant");
+    pub(in crate::envs) fn train_input(&self, train: &Train) -> Option<Input> {
+        let consist = self.consists.get(train)?;
+        let constraints = self.constraints.get(train)?;
+        Some(Input(consist.clone(), constraints.clone()))
+    }
+}
+
+/// Internal context allowing running pathfinding tasks
+///
+/// Built from a [PathfindingEnv] using [Runner::new] which builds indexes in
+/// its internal state. Do not create this struct directly.
+pub(in crate::envs) struct Runner<Train>
+where
+    Train: Clone + Hash + Eq + Send + Sync + 'static,
+{
+    core_env: CoreEnv,
+    pathfinding_inputs: Arc<PathfindingEnvInputs<Train>>,
+    input_index: DashMap<Input, TrainSet<Train>>,
+}
+
+impl<Train> Runner<Train>
+where
+    Train: Clone + Hash + Eq + Send + Sync + 'static,
+{
+    pub(in crate::envs) fn new(env: PathfindingEnv<Train>) -> Self {
+        let (core_env, pathfinding_inputs) = env.decompose();
+        let input_index = DashMap::<Input, TrainSet<Train>>::new();
+        for train in pathfinding_inputs.all_trains() {
+            let consist = pathfinding_inputs
+                .consists
+                .get(&train)
+                .expect("all_trains invariant");
+            let constraints = pathfinding_inputs
+                .constraints
+                .get(&train)
+                .expect("all_trains invariant");
             let input = Input(consist.clone(), constraints.clone());
-            self.trains.entry(input).or_default().insert(train);
+            input_index.entry(input).or_default().insert(train);
+        }
+        Self {
+            core_env: core_env.clone(),
+            pathfinding_inputs,
+            input_index,
         }
     }
 
-    fn train_input(&self, train: &Train) -> Option<Input> {
-        let consist = self.consists.get(train)?;
-        let constraints = self.constraints.get(train)?;
-        Some(Input(consist.clone(), constraints.clone()))
+    pub(in crate::envs) fn input_train_set(&self, input: &Input) -> TrainSet<Train> {
+        self.input_index
+            .get(input)
+            .expect("Runner::new invariant")
```

Comment:
Is this true because the lifetime of `&self` is the same or longer than `&Input`? If yes, we may be able to even express it to make it clear (I'm not entirely sure of my hypothesis, or even of my proposition, since bringing back lifetime when you don't need them... mouarf). Maybe this is just about making the comment in the expect more explicit about that fact?

```rust
    pub(in crate::envs) fn input_train_set<'s, 'i: 's>(&'s self, input: &'i Input) -> TrainSet<Train> {
```